orchagent/mplsrouteorch.cpp

```diff
@@ -421,97 +475,97 @@ bool RouteOrch::addLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey
+                {
+                    next_hop_id = m_neighOrch->getNextHopId(nexthop);
+                }
+                /* See if there is an IP neighbor nexthop */
+                else if (nexthop.label_stack.getSize() &&
+                         m_neighOrch->hasNextHop(NextHopKey(nexthop.ip_address, nexthop.alias)))
+                {
+                    m_neighOrch->addNextHop(nexthop);
+                    next_hop_id = m_neighOrch->getNextHopId(nexthop);
+                }
+                else
+                {
+                    SWSS_LOG_INFO("Failed to get next hop %s for %u",
+                            nextHops.to_string().c_str(), label);
+                    return false;
+                }
             }
         }
-    }
-    /* The route is pointing to a next hop group */
-    else
-    {
-        /* Check if there is already an existing next hop group */
-        if (!hasNextHopGroup(nextHops))
+        /* The route is pointing to a next hop group */
+        else
         {
-            /* Try to create a new next hop group */
-            if (!addNextHopGroup(nextHops))
+            /* Check if there is already an existing next hop group */
+            if (!hasNextHopGroup(nextHops))
             {
-                for (auto it = nextHops.getNextHops().begin(); it != nextHops.getNextHops().end(); ++it)
+                /* Try to create a new next hop group */
+                if (!addNextHopGroup(nextHops))
                 {
-                    const NextHopKey& nextHop = *it;
-                    if (!m_neighOrch->hasNextHop(nextHop))
-                    {
-                        SWSS_LOG_INFO("Failed to get next hop %s in %s, resolving neighbor",
-                                      nextHop.to_string().c_str(), nextHops.to_string().c_str());
-                        m_neighOrch->resolveNeighbor(nextHop);
-                    }
-                }
+                    /* Failed to create the next hop group and check if a temporary route is needed */
 
-                /* Failed to create the next hop group and check if a temporary route is needed */
-
-                /* If the current next hop is part of the next hop group to sync,
-                 * then return false and no need to add another temporary route. */
-                if (it_route != m_syncdLabelRoutes.at(vrf_id).end() && it_route->second.getSize() == 1)
-                {
-                    const NextHopKey& nexthop = *it_route->second.getNextHops().begin();
-                    if (nextHops.contains(nexthop))
+                    /* If the current next hop is part of the next hop group to sync,
+                     * then return false and no need to add another temporary route. */
+                    if (it_route != m_syncdLabelRoutes.at(vrf_id).end() && it_route->second.nhg_key.getSize() == 1)
                     {
-                        return false;
+                        NextHopKey nexthop(it_route->second.nhg_key.to_string());
```

Comment:
Is it no longer possible to extract NextHopKey directly from it_route using previous logic?
const NextHopKey& nexthop = *it_route->second.getNextHops().begin();
It is better to use the original NextHopKey instance than to create a copy from the string.

Reply:
Fixed